lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -328,144 +309,280 @@ newStakePoolLayer gcStatus nl db@DBLayer {..} restartSyncThread = do
+    | ownerStake < ownerPledge = Coin 0
+    | otherwise
+        = afterFees cost margin
+        $ performanceEstimate `fractionOf` optimalRewards rp s sigma_nonmyopic
+  where
+    s     = ownerStakeRelative
+    sigma = stakeRelative
+    t     = tcoin `proportionTo` (totalStake rp)
+
+    sigma_nonmyopic
+        | isTop      = max (getPercentage sigma + t) (z0 rp)
+        | otherwise  = getPercentage s + t
+
+-- | Subtract fixed and margin fees from a 'Coin'.
+afterFees :: Coin -> Percentage -> Coin -> Coin
+afterFees cost margin x = case x `Coin.subtract` cost of
+    Just y  -> oneMinus margin `percentOf` y
+    Nothing -> Coin 0
+
+-- | Optimal rewards for a stake pool
+-- according to Eq.(2) of Section 5.5.3 in SL-D1.
 --
--- Data Combination functions
+-- > optimalRewards s sigma
 --
+-- NOTE: This computation uses 'Double' internally
+-- and is only suitable for the purpose of ranking,
+-- not for computing actual monetary rewards.
+optimalRewards :: RewardParams -> Percentage -> Rational -> Coin
+optimalRewards params s sigma = Coin . round
+    $ factor * (fromIntegral . unCoin . r $ params)
+  where
+    factor = 1 / (1 + a0_)
+        * ( sigma' + s' * a0_ * (sigma' - s'*(z0_-sigma')/z0_) / z0_ )
+
+    z0_, a0_, sigma', s' :: Double
+    z0_    = fromRational (z0 params)
+    a0_    = fromRational (a0 params)
+    sigma' = min (fromRational sigma) z0_
+    s'     = min (fromRational $ getPercentage s) z0_
+
+-- | The desirabilty of a pool is equal to the member rewards at saturation
+-- IF the owner meets their pledge.
+desirability :: RewardParams -> RewardInfoPool -> Coin
+desirability rp RewardInfoPool{..}
+    = afterFees cost margin
+    $ performanceEstimate `fractionOf` optimalRewards rp ownerStakeRelative (z0 rp)
+
+-- | The saturation of a pool is the ratio of the current pool stake
+-- to the fully saturated stake.
+poolSaturation :: RewardParams -> RewardInfoPool -> Double
+poolSaturation rp RewardInfoPool{stakeRelative}
+    = fromRational (getPercentage stakeRelative) / fromRational (z0 rp)
+
+data PoolScore = PoolScore
+    { _desirability :: Coin
+    , _nonMyopicMemberReward :: Coin
+    }
 
--- | Stake pool-related data that has been read from the node using a local
---   state query.
-data PoolLsqData = PoolLsqData
-    { nonMyopicMemberRewards :: Coin
-    , relativeStake :: Percentage
-    , saturation :: Double
-    } deriving (Eq, Show, Generic)
+-- | Compute the desirability and non-myopic rewards for all pools.
+--
+-- To compute the non-myopic rewards, we need to know all pools
+-- in order to rank them by desirability,
+-- and we need to know the stake that the user wants to delegate.
+scorePools
+    :: Ord poolId
+    => RewardParams
+    -> Map poolId (RewardInfoPool, a)
+    -> Coin -- ^ Stake that the user wants to delegate
+    -> Map poolId (PoolScore, RewardInfoPool, a)
+scorePools params pools t
+    = Map.fromList $ zipWith doScore sortedByDesirability areTop
```

Comment:
We should first sort by non-myopic rewards, then sort by desirability, and then sort by seeded random order.

Reply:
Ok, I will adjust the value to sort by. (The actual sorting is done in the `sortByReward` function, which currently sorts according to non-myopic rewards. The code snippet above is internal to the `scorePools` function.)

Reply:
Actually, it turns out that the non-myopic member rewards will never be equal to exactly zero, as they are essentially equal to the rewards for pool stake = owner stake + user stake. Hence, there will probably never be any ties that are resolved by the additional sorting by desirability score. Nonetheless, I will add that.